ml_peg/analysis/physicality/diatomics/analyse_diatomics.py

```diff
@@ -0,0 +1,403 @@
+) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Sort and align energy/force series for a diatomic pair.
+
+    Parameters
+    ----------
+    pair_dataframe
+        Pair-specific dataframe.
+
+    Returns
+    -------
+    tuple[np.ndarray, np.ndarray, np.ndarray]
+        Distances, shifted energies, and projected forces sorted by decreasing
+        distance.
+    """
+    df_sorted = pair_dataframe.sort_values("distance").drop_duplicates("distance")
+    series = df_sorted[["distance", "energy", "force_parallel"]].to_numpy()
+    if len(series) < 3:
+        return np.array([]), np.array([]), np.array([])
+
+    distances = series[:, 0]
+    energies = series[:, 1]
+    forces = series[:, 2]
+
+    descending = np.argsort(distances)[::-1]
+    distances = distances[descending]
+    energies = energies[descending]
+    forces = forces[descending]
+
+    shifted_energies = energies - energies[0]
+    return distances, shifted_energies, forces
+
+
+def count_sign_changes(array: np.ndarray, tol: float) -> int:
+    """
+    Count sign changes in a sequence while ignoring small magnitudes.
+
+    Parameters
+    ----------
+    array
+        Input values.
+    tol
+        Absolute tolerance below which values are treated as zero.
+
+    Returns
+    -------
+    int
+        Number of sign changes exceeding the specified tolerance.
+    """
+    if array.size < 3:
+        return 0
+    clipped = array.copy()
+    clipped = clipped[np.abs(clipped) > tol]
+    signs = np.sign(clipped)
+    sign_flips = signs[:-1] != signs[1:]
+    return int(np.sum(sign_flips))
+
+
+def compute_pair_metrics(
+    df_pair: pd.DataFrame,
+) -> tuple[dict[str, float], float] | tuple[None, None]:
+    """
+    Compute diagnostics for a single diatomic pair.
+
+    Parameters
+    ----------
+    df_pair
+        Pair-specific dataframe.
+
+    Returns
+    -------
+    tuple[dict[str, float], float] | tuple[None, None]
+        Dictionary of metrics and the homonuclear well depth (if applicable).
+    """
+    distances, shifted_energies, projected_forces = prepare_pair_series(df_pair)
+    if distances.size < 3:
+        return None, None
+
+    energy_gradient = np.gradient(shifted_energies, distances)
+    energy_curvature = np.gradient(energy_gradient, distances)
+
+    minima = 0
+    if shifted_energies.size >= 3:
+        second_diff = np.diff(np.sign(np.diff(shifted_energies)))
+        minima = int(np.sum(second_diff > 0))
+
+    inflections = count_sign_changes(energy_curvature, tol=0.5)
+
+    rounded_forces = projected_forces.copy()
+    rounded_forces[np.abs(rounded_forces) < 1e-2] = 0.0
+    force_signs = np.sign(rounded_forces)
+    nonzero_mask = force_signs != 0
+    force_flip_count = (
+        int(np.sum(np.diff(force_signs[nonzero_mask]) != 0))
+        if nonzero_mask.any()
+        else 0
+    )
+
+    well_depth = float(shifted_energies.min())
+
+    spearman_repulsion = np.nan
+    spearman_attraction = np.nan
```

Comment:
```suggestion
    spearman_repulsion = None
    spearman_attraction = None
```
To be consistent with elsewhere? Or does this need to be `np.nan`/ is this dealt with later?